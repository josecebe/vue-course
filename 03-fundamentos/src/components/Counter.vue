<template>
    <h2>{{title}}</h2>
    <p>{{counter}} <sup>2</sup> = {{getSquareValue}}</p>
    <p data-test-id="counter">{{counter}}</p>

    <div class="buttons">
        <button data-test-id="increase-counter-button" @click="increaseCounter">+1</button>
        <button data-test-id="decrease-counter-button" @click="decreaseCounter">-1</button>
    </div>
</template>

<script>
export default {
    name: 'Counter',

    props: {
        title: {
            type: String,
            default: 'Counter',
        },
        start: {
            type: Number,
            default: 10,
            validator(value) {
                return value >= 0;
            },
        },
    },

    data() {
        return {
            counter: this.start,
        }
    },

    methods: {
        increaseCounter() {
            this.counter++;
        },
        decreaseCounter() {
            this.counter--;
        }
    },

    computed: {
        getSquareValue() {
            return this.counter * this.counter;
        },
    }
}
</script>

<style>
button {
    background-color: #64BB87;
    border-radius: 5px;
    border: 0;
    color: white;
    margin : 0 5px;
    padding: 2px 8px;
    transition: background-color 0.3s;
}

button:hover {
    cursor: pointer;
    background-color: #51946c;
}
</style>
