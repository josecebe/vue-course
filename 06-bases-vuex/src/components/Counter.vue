<template>
    <h1>Counter using Vuex</h1>
    <h2>Direct access: {{$store.state.counter.count}}</h2>
    <h2>Computed: {{countComputed}}</h2>
    <h2>Map access: {{count}}</h2>
    <h2>Last mutation: {{lastMutation}}</h2>
    <h2>Last int: {{lastInt}}</h2>

    <button @click="increment">Increment +1</button>
    <button @click="incrementBy(5)">Increment +5</button>
    <button @click="incrementRandomInt" :disabled="isLoading">Increment +RANDOM</button>

    <h2>Direct getter: {{$store.getters['counter/getSquareCount']}}</h2>
</template>

<script>
import {mapState, mapActions} from 'vuex';

export default {
    methods: {
        increment() {
            this.$store.commit('counter/increment');
        },
        incrementBy(val) {
            this.$store.commit('counter/incrementBy', val);
        },
        ...mapActions('counter', ['incrementRandomInt'])
        // ...mapActions('counter', {
        //     randomInt: 'incrementRandomInt'
        // }),
    },

    computed: {
        countComputed() {
            return this.$store.state.counter.count;
        },
        ...mapState('counter', ['count','lastMutation','lastInt','isLoading'])
        // ...mapState({ 
        //     count: state => state.counter.count,
        //     lastMutation: 'lastMutation'
        // }),
    }
}
</script>
