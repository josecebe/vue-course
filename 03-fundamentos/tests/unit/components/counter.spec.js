import {shallowMount} from '@vue/test-utils';
import Counter from '@/components/Counter.vue';

describe('Counter Test', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(Counter);
    });

    test('Component should match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('H2 should contain the default value', () => {
        const h2Elem = wrapper.find('h2');
        expect(h2Elem.exists()).toBeTruthy();

        const h2Text = h2Elem.text();
        expect(h2Text).toBe('Counter');
    });

    test('Default value should be 10', async() => {
        const pText = wrapper.find('[data-test-id="counter"]').text();
        expect(pText).toBe('10');
    });

    test('Increase & decrease buttons are increasing/decreasing 1 to counter', async() => {
        const [increaseBtn, decreaseBtn] = wrapper.findAll('[data-test-id="increase-counter-button"], [data-test-id="decrease-counter-button"]');

        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');
        await decreaseBtn.trigger('click');
        await decreaseBtn.trigger('click');

        const counterValue = wrapper.find('[data-test-id="counter"]').text();
        expect(counterValue).toBe('11');
    });

    test('Default "start" value is working correctly', () => {
        const {start} = wrapper.props();
        const counterValue = Number(wrapper.find('[data-test-id="counter"]').text());
        expect(counterValue).toBe(start);
    });
});
