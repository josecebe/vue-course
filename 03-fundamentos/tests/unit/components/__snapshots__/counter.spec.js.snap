// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Test Component should match snapshot 1`] = `
<h2>Counter</h2>
<p>10 <sup>2</sup> = 100</p>
<p data-test-id="counter">10</p>
<div class="buttons"><button data-test-id="increase-counter-button">+1</button><button data-test-id="decrease-counter-button">-1</button></div>
`;
